server/src/test/java/org/elasticsearch/common/geo/GeometryTreeTests.java

```diff
@@ -217,4 +222,53 @@ public void testPacManPoints() throws Exception {
         GeometryTreeReader reader = new GeometryTreeReader(output.bytes().toBytesRef(), TestCoordinateEncoder.INSTANCE);
         assertTrue(reader.intersects(Extent.fromPoints(xMin, yMin, xMax, yMax)));
     }
+
+    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/37206")
+    public void testRandomGeometryIntersection() throws IOException {
+        int testPointCount = randomIntBetween(100, 200);
+        Point[] testPoints = new Point[testPointCount];
+        double extentSize = randomDoubleBetween(1, 10, true);
+        boolean[] intersects = new boolean[testPointCount];
+        for (int i = 0; i < testPoints.length; i++) {
+            testPoints[i] = randomPoint(false);
+        }
+
+        Geometry geometry = randomGeometryTreeGeometry();
+        GeoShapeIndexer indexer = new GeoShapeIndexer(true, "test");
+        geometry = indexer.prepareForIndexing(geometry);
+
+        for (int i = 0; i < testPointCount; i++) {
+            int cur = i;
+            intersects[cur] = fold(geometry, false, (g, s) -> s || intersects(g, testPoints[cur], extentSize));
+        }
+
+        for (int i = 0; i < testPointCount; i++) {
+            assertEquals(intersects[i], intersects(geometry, testPoints[i], extentSize));
+        }
+    }
+
+    private boolean intersects(Geometry g, Point p, double extentSize) throws IOException {
+        // TODO: Make this independent from GeometryTree
+        GeometryTreeWriter writer = new GeometryTreeWriter(g, GeoShapeCoordinateEncoder.INSTANCE);
+        BytesStreamOutput output = new BytesStreamOutput();
+        writer.writeTo(output);
+        output.close();
+        int xMin = GeoShapeCoordinateEncoder.INSTANCE.encodeX(Math.max(p.getX() - extentSize, -180.0));
+        int xMax = GeoShapeCoordinateEncoder.INSTANCE.encodeX(Math.min(p.getX() + extentSize, 180.0));
+        int yMin = GeoShapeCoordinateEncoder.INSTANCE.encodeY(Math.max(p.getY() - extentSize, -90));
+        int yMax = GeoShapeCoordinateEncoder.INSTANCE.encodeY(Math.min(p.getY() + extentSize, 90));
```

Comment:
do you think it would be OK to change this to:

```
int xMin = GeoShapeCoordinateEncoder.INSTANCE.encodeX(Math.max(p.getX() - extentSize, -179.9999));
int xMax = GeoShapeCoordinateEncoder.INSTANCE.encodeX(Math.min(p.getX() + extentSize, 179.9999));
int yMin = GeoShapeCoordinateEncoder.INSTANCE.encodeY(Math.max(p.getY() - extentSize, -89.9999));
int yMax = GeoShapeCoordinateEncoder.INSTANCE.encodeY(Math.min(p.getY() + extentSize, 89.9999));
```

and also fix the other edge-tree reader bug by adding:

```
            // does rectangle's edges intersect or reside inside polygon's edge
            if (extent.minX() <= Math.min(root.x1, root.x2) && extent.minY() <= Math.min(root.y1, root.y2)
                && extent.maxX() >= Math.max(root.x1, root.x2) && extent.maxY() >= Math.max(root.y1, root.y2)) {
                return true;
            }
```

to EdgeTreeReader#crosses on line 204, after the `// does rectangle's edges intersect or reside inside polygon's edge` check?

Reply:
nevermind. Igor and I spoke out-of-github and concluded that it is better to do this as a follow-up